bindings/rust/s2n-tls/src/renegotiate.rs

```diff
@@ -0,0 +1,677 @@
+        }
+    }
+
+    // s2n-tls doesn't support sending client hello requests.
+    // This makes it impossible to test renegotiation without direct access to
+    // s2n-tls internals like the methods for sending arbitrary records.
+    // Instead, we need to use openssl as our server.
+    struct RenegotiateTestPair {
+        client: Connection,
+        server: SslStream<ServerTestStream>,
+    }
+
+    impl RenegotiateTestPair {
+        fn from(mut builder: config::Builder) -> Result<Self, Box<dyn Error>> {
+            // openssl and s2n-tls must be configured to accept each other's
+            // certificates. Some tests will require client auth.
+            //
+            // openssl also requires a properly configured CA cert, which the
+            // default TestPair does not include.
+            let certs_dir = concat!(
+                env!("CARGO_MANIFEST_DIR"),
+                "/../../../tests/pems/permutations/rsae_pkcs_4096_sha384/"
+            );
+            let certs = CertKeyPair::from(certs_dir, "server-chain", "server-key", "ca-cert");
+
+            // Build the s2n-tls client.
+            builder.load_pem(certs.cert(), certs.key())?;
+            builder.trust_pem(certs.cert())?;
+            builder.set_verify_host_callback(InsecureAcceptAllCertificatesHandler {})?;
+            let config = builder.build()?;
+            let s2n_pair = TestPair::from_config(&config);
+            let client = s2n_pair.client;
+
+            // Build the openssl server.
+            let mut ctx_builder = SslContext::builder(SslMethod::tls_server())?;
+            ctx_builder.set_max_proto_version(Some(SslVersion::TLS1_2))?;
+            ctx_builder.set_min_proto_version(Some(SslVersion::TLS1_2))?;
+            ctx_builder.set_certificate_chain_file(certs.cert_path())?;
+            ctx_builder.set_private_key_file(certs.key_path(), SslFiletype::PEM)?;
+            ctx_builder.set_ca_file(certs.ca_path())?;
+            ctx_builder.set_verify(SslVerifyMode::PEER);
+            let openssl_ctx = ctx_builder.build();
+            let openssl_ssl = Ssl::new(&openssl_ctx)?;
+
+            // Connect the openssl server to the same IO that the s2n-tls
+            // client was constructed to use.
+            let server_stream = ServerTestStream(s2n_pair.io);
+            let server = SslStream::new(openssl_ssl, server_stream)?;
+
+            Ok(Self { client, server })
+        }
+
+        // Translate the output of openssl's `accept` to match s2n-tls's `poll_negotiate`.
+        pub fn poll_openssl_negotiate(
+            server: &mut SslStream<ServerTestStream>,
+        ) -> Poll<Result<(), Box<dyn Error>>> {
+            match server.accept() {
+                Ok(_) => Ready(Ok(())),
+                Err(err) => {
+                    if err.code() == ErrorCode::WANT_READ {
+                        Pending
+                    } else {
+                        Ready(Err(err.into()))
+                    }
+                }
```

Comment:
```suggestion
                Err(err) if err.code() == ErrorCode::WANT_READ => Pending,
                Err(err) => Ready(Err(err.into()))
```
Nit: I think including the `if` as part of the match is a bit more idiomatic? I don't have a strong preference either way 🤷